### ISSUE TITLE
Add an `Entities::iter()` method to iterate over all entities when you only have `&World`

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#6228]
## What problem does this solve or what need does it fill?

The ordinary approach for getting the list of entities (query for `Query<Entity>`) doesn't work in contexts where you only have read-only access to the `World`.

## What solution would you like?

Add an `Entities::iter()` or `World::iter_entities()` method.

## What alternative(s) have you considered?

The following snippet works
```rust
        for archetype in world.archetypes().iter() {
            builder.extract_entities(archetype.entities().iter().copied());
```

Borrowed from @mockersf in https://github.com/bevyengine/bevy/pull/6227/files#diff-717fbf581c75ed1fc65c9755af2bf8a2ddc97c02e89f4044f86efa70a2ea4260L45

## Additional context

It's not entirely clear if `Entities::iter()` can be implemented directly without digging in; I'm not sure if we have all of the information needed there.

**Comment by hymm** [comment]
FYI I was able to implement this on `Entities`, but I prefer #6228 as my method requires filtering out entities with invalid archetypes and creating the `Entity` from the `EntityMeta`.

```rust
impl Entities {
    /// Returns an iterator over entities in valid archetypes
    pub fn iter(&'_ self) -> impl Iterator<Item = Entity> + '_ {
        self.meta
            .iter()
            .enumerate()
            .filter(|(_, meta)| meta.location.archetype_id != ArchetypeId::INVALID)
            .map(|(index, meta)| Entity {
                generation: meta.generation,
                id: index as u32,
            })
    }
}
```

**Comment by mockersf** [comment]
closed by #6242

**Comment by chengts95** [comment]
In 0.17 the `World::iter_entities()` is deprecated and it is redirected to `World::Query::<EntityRef>` again.  However, this redirect reintroduces the issue of  `unable to use read-only access to the World`.  In 0.18,  `World::iter_entities()` seems to be removed. Therefore, I hope people should be aware of this.
